### PR TITLE
fix: remove reflection in forking-printer

### PR DESCRIPTION
**Pull request by hugoduncan**
Remove reflection to improve speed and reduce garbage creation.

- [x] The commits are consistent with our [contribution guidelines](../blob/master/.github/CONTRIBUTING.md)
- [x] You've added tests to cover your change(s)
- [x] All tests are passing
- [x] The new code is not generating reflection warnings
- [] You've updated the README (if adding/changing middleware)

### CODE REVIEW

File: src/cider/nrepl/middleware/out.clj [hunk 1]
```diff
@@ -18,6 +18,8 @@
 
 (declare unsubscribe-session)
 
+(set! *warn-on-reflection* true)
```
Comment (vemv) [inline src/cider/nrepl/middleware/out.clj:21]: We're using Eastwood as a reflection linter (CI-backed), you can remove this while also removing all of this https://github.com/clojure-emacs/cider-nrepl/blob/c3d6b69c98ef4c81529b69fa7624662f2006f545/project.clj#L150-L156

If the fix is indeed OK, Eastwood should be happy with it

**Review by vemv — COMMENTED** [review]
Looking good, I like the main fix - not sure if I would have thought of `cond`ing the proxy myself :)

File: src/cider/nrepl/middleware/out.clj [hunk 2]
```diff
@@ -50,17 +52,42 @@ Please do not inline; they must not be recomputed at runtime."}
...
+                  ;; as `int` and `char[]` aren't supported by proxy.
                   (write
                     ([x]
-                     (.write (original-output type) x)
                      (with-out-binding [printer messages type]
-                       (.write printer x)))
+                       (cond
+                         (string? x)
+                         (do
+                           (.write ^Writer (original-output type) ^String x)
```
Comment (vemv) [inline src/cider/nrepl/middleware/out.clj:63]: Extra request (sorry for the pickiness), please verify that there's coverage for all these branches or add it

Context: touching this area recently resulted in an opaque issue that impacted users.

**Comment by bbatsov** [comment]
I can only add - don't forget to update the changelog. :-) I know this is isn't exactly user-visible, but I know some people would be happy to see there's a bit less reflection warnings.

**Review by vemv — APPROVED** [review]
Neat! Thanks for the iteration 🍻

**Review by vemv — APPROVED** [review]
Neat! Thanks for the iteration 🍻

**Pull request merged**